modules/planning/math/discretized_points_smoothing/fem_pose_deviation_smoother.cc

```diff
@@ -112,38 +128,145 @@ bool FemPoseDeviationSmoother::Optimize(const OsqpSettings& solver_settings) {
 
 void FemPoseDeviationSmoother::CalculateKernel(std::vector<c_float>* P_data,
                                                std::vector<c_int>* P_indices,
-                                               std::vector<c_int>* P_indptr) {}
+                                               std::vector<c_int>* P_indptr) {
+  CHECK_GT(num_of_variables_, 4);
+  // Only upper triangle is filled
+  std::vector<std::vector<std::pair<c_int, c_float>>> columns;
+  columns.resize(num_of_variables_);
+  int col_num = 0;
+
+  for (int col = 0; col < 2; ++col) {
```

Comment:
Better comment how you use the vectors, as there are too many magic numbers in this function. It's just not maintainable.